x/trigger/keeper/event_detector.go

```diff
@@ -0,0 +1,72 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/trigger/types"
+)
+
+// DetectBlockEvents Detects triggers that have been activated by their corresponding events.
+func (k Keeper) DetectBlockEvents(ctx sdk.Context) {
+	triggers := k.detectTransactionEvents(ctx)
+	triggers = append(triggers, k.detectBlockHeightEvents(ctx)...)
+	triggers = append(triggers, k.detectTimeEvents(ctx)...)
+
+	for _, trigger := range triggers {
+		k.UnregisterTrigger(ctx, trigger)
+		k.QueueTrigger(ctx, trigger)
+	}
+}
+
+// DetectTransactionEvents Detects triggers that have been activated by transaction events.
+func (k Keeper) detectTransactionEvents(ctx sdk.Context) (triggers []types.Trigger) {
+	detectedTriggers := map[uint64]bool{}
+	for _, event := range ctx.EventManager().GetABCIEventHistory() {
+		matched := k.getMatchingTriggers(ctx, event.GetType(), func(trigger types.Trigger, triggerEvent types.TriggerEventI) bool {
+			if _, isDetected := detectedTriggers[trigger.Id]; isDetected {
+				return false
+			}
+			txEvent := triggerEvent.(*types.TransactionEvent)
+			detected := txEvent.Equals(event)
+			detectedTriggers[trigger.Id] = detected
+			return detected
+		})
+		triggers = append(triggers, matched...)
+	}
+	return
+}
+
+// DetectBlockHeightEvents Detects triggers that have been activated by block height events.
+func (k Keeper) detectBlockHeightEvents(ctx sdk.Context) (triggers []types.Trigger) {
+	triggers = k.getMatchingTriggers(ctx, types.BlockHeightPrefix, func(_ types.Trigger, triggerEvent types.TriggerEventI) bool {
+		blockHeightEvent := triggerEvent.(*types.BlockHeightEvent)
+		return ctx.BlockHeight() >= int64(blockHeightEvent.GetBlockHeight())
+	})
+	return
+}
+
+// DetectTimeEvents Detects triggers that have been activated by block time events.
+func (k Keeper) detectTimeEvents(ctx sdk.Context) (triggers []types.Trigger) {
+	triggers = k.getMatchingTriggers(ctx, types.BlockTimePrefix, func(_ types.Trigger, triggerEvent types.TriggerEventI) bool {
+		blockTimeEvent := triggerEvent.(*types.BlockTimeEvent)
+		return ctx.BlockTime().UTC().Equal(blockTimeEvent.GetTime().UTC()) || ctx.BlockTime().UTC().After(blockTimeEvent.GetTime().UTC())
+	})
+	return
+}
```

Comment:
I'm not sure how sustainable this is. E.g. if there's 100 events a day for the next 2 years, this is doing a whole lot of processing that it doesn't need to do.

It would be better to store the block time events with their trigger time (epoch) in the key. For example, the key would have the format `<prefix><epoch><trigger id>`, or something along that line (I haven't gotten down to the keys area yet).

Then, when looking for time events that should be triggered, you start the iterator with just `<prefix>` and end the iterator at `<prefix><epoch><max uint64>`. Alternatively, you could use the `<max trigger id>` there instead of `<max uint64>`, but that'd require an extra lookup.

This key format is what's used in the authz module for deleting expired authorizations, and the attribute module for deleting expired attributes.